### PR TITLE
Reworks the use of magnetism and lightning to make them actually work

### PR DESCRIPTION
**Pull request by farie82**
## What Does This PR Do
Makes it so that both the lightning and magnetism spell start charging when clicking on the button. Then once you click on a mob (or near it) they will activate on said mob. Before it was click on button. Click on button again to activate
Clicking the button again will try to stop the spell from charging. If it is over 50% done it will fail to stop.
Makes it so the progress bar and the charge up sound actually both go away when you use or cancel the spell.
Makes magnetism actually work. Be it a crappy spell it works as intended now.
Fixes a bug I accidentally left in with lightning... which also made it unusable.
Refactored both spells to not just be copypasta from each other and instead, they are now a subtype of a spell.

Approved by @dearmochi as a fix/refactor since it was in such a broken state

Magnet sprite by CinnamonSnowball

## Why It's Good For The Game
Fixes two of the most broken/clunky spells in the game.

## Images of changes
Both gifs still have no magnet icon. But it is there I promise
T693sPU0Dv
8YQkGTEYVG

## Changelog
:cl: Farie82 CinnamonSnowball
tweak: Both the lightning and magnetism spell now start charging on the button click and to fire them you have to click on a mob (or near it). Clicking the button again will try and cancel it.
fix: Both the lightning and magnetism spell now remove the progress bar and sound when you cast the spell or stop casting it.
fix: Both the lightning and magnetism spell now actually work (partial whoops from me)
imageadd: The magnet spell now has a sprite icon!
/:cl:

### CODE REVIEW

File: code/__HELPERS/mobs.dm [hunk 1]
```diff
@@ -333,7 +333,7 @@
 
 	msg_admin_attack("[key_name_admin(user)] vs [target_info]: [what_done]", loglevel)
 
-/proc/do_mob(mob/user, mob/target, time = 30, uninterruptible = 0, progress = 1, list/extra_checks = list())
+/proc/do_mob(mob/user, mob/target, time = 30, progress = 1, list/extra_checks = list(), only_use_extra_checks = FALSE)
```
Comment (farie82) [inline code/__HELPERS/mobs.dm:336]: I checked all the uses of `do_mob` and only the magnetism and lightning spell used `uninterruptible`

Comment (hal9000PR) [reply code/__HELPERS/mobs.dm:336]: Might be useful later no?

Comment (farie82) [reply code/__HELPERS/mobs.dm:336]: The functionality still remains. Just have to set `only_use_extra_checks` on `TRUE` and leave the `extra_checks` list empty. Or add your own

**Comment by Denthamos** [comment]
Just verifying, but how does the lightning bolt/magnetism spell handle when you purchase it multiple times/max it out?  Does it charge faster and let you fire it off faster properly?  I remember there was an issue with that in the past too.

**Comment by farie82** [comment]
> Just verifying, but how does the lightning bolt/magnetism spell handle when you purchase it multiple times/max it out? Does it charge faster and let you fire it off faster properly? I remember there was an issue with that in the past too.

I did not find code that would suggest upgrading lowers the charge-up time. From testing, I also found that upgrades only lower the cooldown between uses. Not the actual charge-up time.

File: code/datums/spells/charge_up.dm [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	else
+		to_chat(user, "<span class='danger'>[stop_charging_fail_text]</span>")
+		return FALSE
+
+/obj/effect/proc_holder/spell/charge_up/proc/get_energy_charge()
+	return min(world.time - start_time, max_charge_time)
+
+/obj/effect/proc_holder/spell/charge_up/proc/StartChargeup(mob/user)
+	to_chat(user, "<span class='notice'>[start_charging_text]</span>")
+	user.overlays.Add(charge_up_overlay)
```
Comment (dearmochi) [inline code/datums/spells/charge_up.dm:53]:
```suggestion
	user.add_overlay(charge_up_overlay)
```

File: code/datums/spells/charge_up.dm [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+	if(do_mob(user, user, max_charge_time, extra_checks = list(CALLBACK(src, .proc/stopped_casting)), only_use_extra_checks = TRUE))
+		if(start_time)
+			Discharge(user)
+
+/obj/effect/proc_holder/spell/charge_up/proc/stopped_casting()
+	return start_time == 0
+
+/obj/effect/proc_holder/spell/charge_up/proc/Reset(mob/user)
+	start_time = 0
+	user.overlays.Remove(charge_up_overlay)
```
Comment (dearmochi) [inline code/datums/spells/charge_up.dm:65]:
```suggestion
	user.cut_overlay(charge_up_overlay)
```

**Review by dearmochi — CHANGES_REQUESTED** [review]
Besides this, should be fine

File: code/datums/spells/charge_up.dm (hunk 2 above)
Comment (farie82) [reply code/datums/spells/charge_up.dm]: Good one. Somehow forgot those existed. Tested and works as expected!

**Review by dearmochi — APPROVED** [review]

**Pull request merged**